numerics/root_finders_body.hpp

```diff
@@ -57,6 +61,103 @@ Argument Bisect(Function f,
   }
 }
 
+// The implementation is translated from the ALGOL 60 in [Bre73], chapter 4,
+// section 6, with the notation adjusted to more closely mirror the formulæ from
+// the preceding sections.
+template<typename Argument, typename Function>
+Argument Brent(Function f,
+               Argument const& lower_bound,
+               Argument const& upper_bound) {
+  using Value = decltype(f(lower_bound));
+  Value const zero{};
+
+
+  // We do not use |std::numeric_limits<double>::epsilon()|, because it is 2ϵ in
+  // Brent’s notation: Brent uses ϵ = β^(1-τ) / 2 for rounded arithmetic, see
+  // (2.9).
+  constexpr double ϵ = ScaleB(0.5, 1 - std::numeric_limits<double>::digits);
+
+  Argument a = lower_bound;
+  Argument b = upper_bound;
+  Argument c;
+
+  Difference<Argument> d;
+  Difference<Argument> e;
+
+  Value f_a = f(a);
+  Value f_b = f(b);
+  Value f_c;
+
+  if (f_a == zero) {
+    return a;
+  }
+  if (f_b == zero) {
+    return b;
+  }
+  CHECK_NE(Sign(f_a), Sign(f_b))
+      << "\nlower: " << lower_bound << u8" ↦ " << f_a << ", "
+      << "\nupper: " << upper_bound << u8" ↦ " << f_b;
+
+interpolation:
+  c = a;
+  f_c = f_a;
+  d = e = b - a;
+extrapolation:
+  if (Abs(f_c) < Abs(f_b)) {
+    a = b;
+    b = c;
+    c = a;
+    f_a = f_b;
+    f_b = f_c;
+    f_c = f_a;
+  }
+  Difference<Argument> const δ = 2 * ϵ * Abs(b - Argument{});
+  Difference<Argument> const m = 0.5 * (c - b);
+  if (Abs(m) > δ && f_b != zero) {
+    // See if a bisection is forced.
+    if (Abs(e) < δ || Abs(f_a) <= Abs(f_b)) {
+      d = e = m;
+    } else {
+      Difference<Argument> p;
+      double q;
+      if (a == c) {
+        // Linear interpolation.
+        double const s = f_b / f_a;
+        p = 2 * m * s;
+        q = 1 - s;
+      } else {
+        // Inverse quadratic interpolation.
+        double const r₁ = f_a / f_c;
+        double const r₂ = f_b / f_c;
+        double const r₃ = f_b / f_a;
+        p = r₃ * (2 * m * r₁ * (r₁ - r₂) - (b - a) * (r₂ - 1));
+        q = (r₁ - 1) * (r₂ - 1) * (r₃ - 1);
+      }
+      if (Sign(p).is_positive()) {
+        q = -q;
+      } else {
+        p = -p;
+      }
+      if (2 * p < 3 * m * q - Abs(δ * q) && p < Abs(0.5 * e * q)) {
+        e = d;
+        d = p / q;
+      } else {
+        d = e = m;
+      }
+    }
+    a = b;
+    f_a = f_b;
+    b += Abs(d) > δ ? d : δ * Sign(m);
+    f_b = f(b);
+    if (Sign(f_b) == Sign(f_c)) {
+      goto interpolation;
```

Comment:
I realise that you are trying to stick to the paper, but I am going to claim that two nested loop would read better and, unlike, say, integrators, the control structure is not hard to transform into loops (`break` if the signs match, `return` if the condition at line 116 is false).